ford/sourceform.py

```diff
@@ -1346,55 +1377,55 @@ def strip_type(s):
+            # procs
+            labels.update(getattr(context, "all_procs", {}))
+            # boundprocs
+            labels.update(
+                {bp.name.lower(): bp for bp in getattr(context, "boundprocs", [])}
+            )
+            # types
+            labels.update(getattr(context, "all_types", {}))
+            # extended type
+            extend_type = context
+            while extend_type := getattr(extend_type, "extends", None):
+                labels[extend_type.name.lower()] = extend_type
+            # vars
+            labels.update(getattr(context, "all_vars", {}))
+            # local vars
+            labels.update({a.name.lower(): a for a in getattr(context, "args", [])})
+            if retvar := getattr(context, "retvar", None):
+                labels[retvar.name.lower()] = retvar
+            labels.update(
+                {v.name.lower(): v for v in getattr(context, "variables", [])}
+            )
 
-        # traverse the call chain
-        for c in [c.lower() for c in call.chain[1:]]:
-            new_call_type = None
-            # try call type is a variable
-            if hasattr(call_type, "variables"):
-                new_call_type_str = None
-                for v in call_type.variables:
-                    if v.name.lower() == c:
-                        new_call_type_str = strip_type(v.full_type)
-                        break
-                new_call_type = call_type.all_types.get(new_call_type_str, None)
-
-            # not a variable, try call type is an extended type
-            if new_call_type is None and isinstance(call_type, FortranType):
-                extend_type: Optional[FortranType] = call_type
-                while extend_type := getattr(extend_type, "extends", None):
-                    if extend_type.name.lower() == c:
-                        new_call_type = extend_type
-                        break
+            # collect the context from the return of the item matching the call name
+            item = labels.get(call, None)
+            # last item shouldn't be a context
+            if i == len(call_chain) - 1:
+                return item
 
-            # not a subtype, give up
-            if new_call_type is None:
+            if item is None:
+                # failed to find context
+                return None
+            elif hasattr(item, "retvar"):
+                type_str = strip_type(item.retvar.full_type)
+                context = item.all_types.get(type_str, None)
+            elif isinstance(item, FortranType):
+                context = item
+            elif isinstance(item, FortranVariable):
+                type_str = strip_type(item.full_type)
+                context = item.parent.all_types.get(type_str, None)
+            else:
                 return None
 
-            call_type = new_call_type
-
-        return call_type
+            if context is None:
+                # failed to find context
+                return None
```

Comment:
I'm unsure about this. Not finding the context just means that any of the contexts leading up to the call aren't registered in Ford. This could simply be because the context is from an external module, and could be quite common, and expected in some codebases.